Added Features

nickngsr opened the pull request:
Adds

- Loxone DNS
- Download Socket
- Rudimentary version checking
- Keepalive
- New way to build a socket

Fixes some issues relating to the reconnect go routine and also some scenarios where calling Close() would cause hanging.

Commit:
(fix) reconnect loop never stops

Commit:
(fix) handleMessages loop not stopping after calling Close
(fix) readPump hanging if the reconnect handler returns after calling Close

Commit:
(feat) allow configuration of reconnect behaviour
(feat) expose the ability to send an encrypted command
(feat) allow register events to be set on instantiation
(refactor) move where RegisterEvents is called on reconnect to allow it to be a configurable option on new websocket
(refactor) add info log for starting the disconnect loop
(docs) add comments for some exposed functions

Commit:
(feat) recover from reflect panic and provide an error instead

Commit:
(fix) make ComputeHmac256 use SHA256 instead of SHA1 hash
(feat) add ComputeHmac1 for SHA1 hash
(fix) return data instead of empty byte array if no padding bytes on AES decrypt
(test) Add new test to test for salt decryption
(test) Add new test for SHA256 HMAC
(test) Change existing test to make it obvious it is checking SHA1 HMAC
(refactor) fill utils.FakeWebSocket with missing methods
(feat) Check for miniserver version and start enabling/disabling features based on version
(refactor) return a receive only channel from GetEvents
(feat) add GetFile method to download binary files (needs changing to use own socket)
(refactor) remove unnecessary storing of one time salt on key exchange
(fix) expose token fields so mapstructure can populate them
(feat) add IsValid function to token to check expiry date
(feat) extend options interface to authenticate with JWT or username and password
(refactor) remove some unnecessary defers
(feat) introduce locking on websocket to allow messages to be sent from goroutines (can be refactored to a write pump)
(refactor) move getPublicKeyFromServer and create token to websocketImpl struct rather than copying struct values
(feat) now reuses stored token on reconnects instead of generating a new one
(refactor) removed variable/import name collision
(fix) returned the correct HMAC on user hash if SHA256 was specified

Commit:
(refactor) add nosec to false positives with gosec
(docs) add examples for the New function
(docs) add descriptions for all the WebsocketOptions
(refactor) remove uneccessary passing of the uniqueId variable
(feat) automatically get salt if zero string passed to reuseToken

Commit:
(feat) add keepalive mechanism
(feat) implement read timeouts to handle Miniservers falling off the face of the earth
(feat) expose stop channel as receive only to detect closing
(feat) add timeout to http requests

Commit:
(chore) update gorilla websocket library
(refactor) check stop channel is closed instead of using atomic int for reconnect handler check
(feat) add LoxoneDownloadSocket as an interface over the current websocketImpl
(fix) handle Close while in a reconnect loop

Commit:
chore: update go module to 1.17
feat: add tls support
feat: allow for skipping certificate verification
feat: add support for Loxone Cloud DNS
refactor: reduce code duplication for instantiating socket

Commit:
chore: update README.md

XciD commented:
Great Stuff. I will add you as contributor.

Feel free to merge when you are ok

nickngsr commented:
Thanks, happy to help!

Commit:
refactor: remove deprecated ioutil functions
fix: goroutines still running if socket fails on first connect
fix: unnecessary socket close in deferred readPump function panicking if read pump stopped before first socket connect
fix: stop error message if socket closed normally
fix: readPump not returning straight away after disconnect

Commit:
fix: default ports not used if CloudDNS does not return port

Commit:
feat: add isConnected flag to track miniserver connect state

Commit:
feat: make connected flag thread safe as well as add connection event channel

Commit:
fix: fix silly return of nil instead of channel
feat: add ConnectionEvents to download socket

Commit:
refactor: (breaking change) remove connectionEvents and add connected handler

Commit:
refactor: buffer disconnect channel instead of select statement

nickngsr committed:
Merge branch 'master' into master

nickngsr commented:
Sorry for being AWOL for a while, this has been soak tested and ready to merge
